### PR TITLE
fix: prevent broken links for flattened modules

### PR DESCRIPTION
**Pull request by lorenzolewis**
**Describe the pull request**

Using a config shown in `astro.module.config.ts` will result in a namespaces section in `index.md` like this:

```md
- [bar](/api-module//namespacebar/)
- [foo](/api-module//namespacefoo/)
- [functions](/api-module//namespacefunctions/)
- [shared](/api-module//namespaceshared/)
- [types](/api-module//namespacetypes/)
```
(notice the extra slash)

**Why**

Broken links 

**How**

This re-works the link generation to account for if there are NO valid segments in a URL (which was causing this extra `/`).

**Screenshots**

N/A

**Note**: This one is based on https://github.com/HiDeoo/starlight-typedoc/pull/10 and includes the changes in there since it would cause a merge conflict once #10 is merged anyway

### CODE REVIEW

File: example/astro.module.config.ts
Comment (HiDeoo) [inline example/astro.module.config.ts:1]: I think we should only add new Starlight setup when they're required by Playwright tests.

File: example/package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   "scripts": {
     "dev": "astro dev",
     "dev:single-entrypoints": "astro dev --config astro.config.ts",
+    "dev:module": "astro dev --config astro.module.config.ts",
```
Comment (HiDeoo) [inline example/package.json:11]:
```suggestion
```

If we remove the `astro.module.config.ts` example, we can also remove this script.

File: example/src/content/docs/.gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 # Autogenerated by https://github.com/HiDeoo/starlight-typedoc
 api/
+api-module/
```
Comment (HiDeoo) [inline example/src/content/docs/.gitignore:3]:
```suggestion
```

Same goes for this one.

File: packages/starlight-typedoc/tests/unit/typedoc.test.ts [hunk 3]
```diff
@@ -91,3 +91,69 @@ test('should not add `README.md` module files for multiple entry points', async
   expect(writeFileSyncSpy).toHaveBeenCalled()
   expect(filePaths.some((filePath) => filePath.endsWith('README.md'))).toBe(false)
 })
+
+test('should support overriding typedoc-plugin-markdown readme and index page generation', async () => {
+  await generateTypeDoc({
+    ...starlightTypeDocOptions,
+    typeDoc: {
+      readme: 'README.md',
```
Comment (HiDeoo) [inline packages/starlight-typedoc/tests/unit/typedoc.test.ts]:
```suggestion
      ...starlightTypeDocOptions.typeDoc,
      readme: 'README.md',
```

Missed that in the previous PR but it avoid spamming the console during tests with useless logs. Might do a helper later to avoid passing that in explicitly but I'll wait to see if I want to refactor the tests at some point.

File: packages/starlight-typedoc/tests/unit/typedoc.test.ts [hunk 4]
```diff
@@ -91,3 +91,69 @@ test('should not add `README.md` module files for multiple entry points', async
...
+  const writeFileSyncSpy = vi.mocked(fs.writeFileSync)
+  const filePaths = writeFileSyncSpy.mock.calls.map((call) => call[0].toString())
+
+  expect(filePaths.some((filePath) => filePath.endsWith('index.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.bar.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.foo.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.functions.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.shared.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.types.md'))).toBe(true)
+})
```
Comment (HiDeoo) [inline packages/starlight-typedoc/tests/unit/typedoc.test.ts:133]: Is this test relevant to this PR? I may be wrong but I think the changes in this PR won't change the output of this test?

File: packages/starlight-typedoc/tests/unit/typedoc.test.ts [hunk 5]
```diff
@@ -91,3 +91,69 @@ test('should not add `README.md` module files for multiple entry points', async
...
+      outputFileStrategy: 'modules',
+      entryFileName: 'index.md',
+      skipIndexPage: false,
+      flattenOutputFiles: true,
+    },
+    entryPoints: ['../../fixtures/src/module.ts'],
+  })
+
+  const writeFileSyncSpy = vi.mocked(fs.writeFileSync)
+  const [, indexString] = writeFileSyncSpy.mock.calls.find((call) => call[0].toString().endsWith('index.md')) as [
```
Comment (HiDeoo) [inline packages/starlight-typedoc/tests/unit/typedoc.test.ts:148]:
```suggestion
  const [, content] = writeFileSyncSpy.mock.calls.find((call) => call[0].toString().endsWith('index.md')) as [
```

Nit picking but I think it makes a bit more sense, at least to me.

File: packages/starlight-typedoc/tests/unit/typedoc.test.ts [hunk 6]
```diff
@@ -91,3 +91,69 @@ test('should not add `README.md` module files for multiple entry points', async
...
+    fs.PathOrFileDescriptor,
+    string
+  ]
+  const indexContents = indexString.split('\n')
+
+  expect(indexContents.find((line) => line.startsWith('- [bar]'))).toBe('- [bar](/api/namespacebar/)')
+  expect(indexContents.find((line) => line.startsWith('- [foo]'))).toBe('- [foo](/api/namespacefoo/)')
+  expect(indexContents.find((line) => line.startsWith('- [functions]'))).toBe('- [functions](/api/namespacefunctions/)')
+  expect(indexContents.find((line) => line.startsWith('- [shared]'))).toBe('- [shared](/api/namespaceshared/)')
+  expect(indexContents.find((line) => line.startsWith('- [types]'))).toBe('- [types](/api/namespacetypes/)')
```
Comment (HiDeoo) [inline packages/starlight-typedoc/tests/unit/typedoc.test.ts:158]:
```suggestion

  expect(
    content.includes(`
- [bar](/api/namespacebar/)
- [foo](/api/namespacefoo/)
- [functions](/api/namespacefunctions/)
- [shared](/api/namespaceshared/)
- [types](/api/namespacetypes/)`)
  ).toBe(true)
```

At this point, I wonder if it's just not easier to write the markdown we expect. Any thoughts?

If we end up querying markdown a bit more in the future, I might as well add a helper to parse the markdown and easily query various type of elements in it.

File: packages/starlight-typedoc/tests/unit/typedoc.test.ts [hunk 7]
```diff
@@ -91,3 +91,69 @@ test('should not add `README.md` module files for multiple entry points', async
...
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.functions.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.shared.md'))).toBe(true)
+  expect(filePaths.some((filePath) => filePath.endsWith('Namespace.types.md'))).toBe(true)
+})
+
+test('should output index with correct module path', async () => {
+  await generateTypeDoc({
+    ...starlightTypeDocOptions,
+    typeDoc: {
+      outputFileStrategy: 'modules',
```
Comment (HiDeoo) [inline packages/starlight-typedoc/tests/unit/typedoc.test.ts:139]:
```suggestion
      ...starlightTypeDocOptions.typeDoc,
      outputFileStrategy: 'modules',
```

Same as above

**Review by HiDeoo — CHANGES_REQUESTED** [review]
Nice changes, got some small questions/tweaks but overall I think it's great and something that we should merge asap.

**Comment by HiDeoo** [comment]
Great GitHub UI not notifying me during the review some changes were pushed 😓 

I think the review should still be valid and not messed up but if you have any question, let me know.

File: packages/starlight-typedoc/tests/unit/typedoc.test.ts (hunk 4 above)
Comment (lorenzolewis) [reply packages/starlight-typedoc/tests/unit/typedoc.test.ts:133]: It's not strictly for anything here, it was one I just added in to check the files were generated as needed (and tbh to first get my head wrapped around vitest a little better).

But this is something that's all handled up stream and not here so can strip them out if wanted, didn't really think about it before PR-ing.

**Comment by lorenzolewis** [comment]
That should be all those fixes in, lmk if you'd like me to pull that extra test out.

**Comment by lorenzolewis** [comment]
Absolutely, any chance to up my testing game 💪

**Comment by HiDeoo** [comment]
There we go, [`v0.3.2`](https://github.com/HiDeoo/starlight-typedoc/releases/tag/v0.3.2) has been published to npm.

Thanks again for your contributions 🥳
